src/spox/_attributes.py

```diff
@@ -188,37 +202,98 @@ def maybe(
 class AttrFloat32s(_AttrIterable[float]):
     _attribute_proto_type_int = AttributeProto.FLOATS
 
-    def _to_onnx_deref(self, key: str) -> AttributeProto:
+    def _to_onnx_deref(self, key: str, only_typecheck: bool = False) -> AttributeProto:
         # ensure values are all floats
         values = [float(v) for v in self.value]
-        return make_attribute(key, values, attr_type=AttributeProto.FLOATS)
+        return (
+            _make_attribute_type(values)
+            if only_typecheck
+            else make_attribute(key, values, attr_type=AttributeProto.FLOATS)
+        )
 
 
 class AttrInt64s(_AttrIterable[int]):
     _attribute_proto_type_int = AttributeProto.INTS
 
-    def _to_onnx_deref(self, key: str) -> AttributeProto:
-        return make_attribute(key, self.value, attr_type=AttributeProto.INTS)
+    def _to_onnx_deref(self, key: str, only_typecheck: bool = False) -> AttributeProto:
+        return (
+            _make_attribute_type(self.value)
+            if only_typecheck
+            else make_attribute(key, self.value, attr_type=AttributeProto.INTS)
+        )
 
 
 class AttrStrings(_AttrIterable[str]):
     _attribute_proto_type_int = AttributeProto.STRINGS
 
-    def _to_onnx_deref(self, key: str) -> AttributeProto:
-        return make_attribute(
-            key, [v.encode() for v in self.value], attr_type=AttributeProto.STRINGS
+    def _to_onnx_deref(self, key: str, only_typecheck: bool = False) -> AttributeProto:
+        return (
+            _make_attribute_type(self.value)
+            if only_typecheck
+            else make_attribute(key, self.value, attr_type=AttributeProto.STRINGS)
         )
 
 
 class AttrTensors(_AttrIterable[np.ndarray]):
     _attribute_proto_type_int = AttributeProto.TENSORS
 
-    def _to_onnx_deref(self, key: str) -> AttributeProto:
+    def _to_onnx_deref(self, key: str, only_typecheck: bool = False) -> AttributeProto:
         tensors = [from_array(t) for t in self.value]
-        return make_attribute(key, tensors, attr_type=AttributeProto.TENSORS)
+        return (
+            _make_attribute_type(tensors)
+            if only_typecheck
+            else make_attribute(key, tensors, attr_type=AttributeProto.TENSORS)
+        )
 
 
 def _deref(ref: _Ref[T]) -> T:
     if isinstance(ref._concrete._value, _Ref):
         return _deref(ref._concrete._value)
     return ref._concrete._value
+
+
+def _make_attribute_type(value: Any) -> AttributeProto:
+    return AttributeProto(type=_deduce_type(value))
+
+
+# With some simplifying modifications, this is a shameless copy from
+# https://github.com/onnx/onnx/blob/b60f69412abb5393ab819b936b473f83867f6c87/onnx/helper.py#L838
+# TODO in this PR: we should probably abstract this in onnx and reuse it in onnx's make_attribute; open an issue or PR
```

Comment:
This is indeed an uncomfortable situation and I’m not sure what’s a good way about it. Validation constructing a throwaway protobuf til now was a lazy solution that did the job, but is as turns out too inefficient. Maybe there’s other things to piggy-back off of (numpy?)? Perhaps at least the code could be rewritten in a better style rather than pasted.

Reply:
I'm also unsure what the best approach is. Ideally, we would modularize this upstream, and then reuse the type checking component. Is this realistic?

>  Maybe there’s other things to piggy-back off of (numpy?)

Could you elaborate? We want exactly the same logic as in `make_attribute`, and since the latter is not piggybacking on anything else, I don't see how we can.

> Perhaps at least the code could be rewritten in a better style rather than pasted.

The closer we are to the original implementation, the better. But if you have any concrete suggestions, happy to reflect them.

Reply:
> I'm also unsure what the best approach is. Ideally, we would modularize this upstream, and then reuse the type checking component. Is this realistic?

Yes, I don’t see issues if one makes the right argument for it upstream. We will have to wait for the next ONNX release, though.

> Could you elaborate? We want exactly the same logic as in make_attribute, and since the latter is not piggybacking on anything else, I don't see how we can.

Yeah, just a random idea - I guess it doesn’t really work out.

> The closer we are to the original implementation, the better. But if you have any concrete suggestions, happy to reflect them.

I would replace the if-statements with a loop (similar as for iterables), and make the raised exceptions more consistent (why are some ValueErrors?). Things like `len(value) == 0` for a list are stylistic, but hurt my soul.

Reply:
We removed this part (as we won't try this optimisation yet), so I am resolving this conversation.